agent/routine-leak-checker/leak_test.go

```diff
@@ -0,0 +1,117 @@
+package leakcheck
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/fortytw2/leaktest"
+	"github.com/hashicorp/consul/agent"
+	"github.com/hashicorp/consul/sdk/freeport"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/hashicorp/consul/testrpc"
+	"github.com/hashicorp/consul/tlsutil"
+	"github.com/stretchr/testify/require"
+)
+
+func testTLSCertificates(serverName string) (cert string, key string, cacert string, err error) {
+	// generate CA
+	serial, err := tlsutil.GenerateSerialNumber()
+	if err != nil {
+		return "", "", "", err
+	}
+	signer, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return "", "", "", err
+	}
+	ca, err := tlsutil.GenerateCA(signer, serial, 365, nil)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	// generate leaf
+	serial, err = tlsutil.GenerateSerialNumber()
+	if err != nil {
+		return "", "", "", err
+	}
+
+	cert, privateKey, err := tlsutil.GenerateCert(
+		signer,
+		ca,
+		serial,
+		"Test Cert Name",
+		365,
+		[]string{serverName},
+		nil,
+		[]x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+	)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	return cert, privateKey, ca, nil
+}
+
+func setupPrimaryServer(t *testing.T) *agent.TestAgent {
+	d := testutil.TempDir(t, "leaks-primary-server")
+	t.Cleanup(func() { os.RemoveAll(d) })
+
+	certPEM, keyPEM, caPEM, err := testTLSCertificates("server.primary.consul")
+	require.NoError(t, err)
+
+	certPath := filepath.Join(d, "cert.pem")
+	keyPath := filepath.Join(d, "key.pem")
+	caPath := filepath.Join(d, "cacert.pem")
+
+	require.NoError(t, ioutil.WriteFile(certPath, []byte(certPEM), 0600))
+	require.NoError(t, ioutil.WriteFile(keyPath, []byte(keyPEM), 0600))
+	require.NoError(t, ioutil.WriteFile(caPath, []byte(caPEM), 0600))
+
+	aclParams := agent.DefaulTestACLConfigParams()
+	aclParams.PrimaryDatacenter = "primary"
+	aclParams.EnableTokenReplication = true
+
+	config := `
+	   server = true
+		datacenter = "primary"
+		primary_datacenter = "primary"
+		enable_central_service_config = true
+		
+		connect {
+			enabled = true
+		}
+		
+		auto_encrypt {
+			allow_tls = true
+		}
+	` + agent.TestACLConfigWithParams(aclParams)
+
+	a := agent.NewTestAgent(t, config)
+	t.Cleanup(func() { a.Shutdown() })
+
+	testrpc.WaitForTestAgent(t, a.RPC, "primary", testrpc.WithToken(agent.TestDefaultMasterToken))
+
+	return a
+}
+
+func TestTestAgentLeaks_Server(t *testing.T) {
```

Comment:
I wonder how reliable this will be sitting in a separate package. I imagine it is going to be hard to keep this synchronized with all the new features and changes we make.

What do you think of https://godoc.org/go.uber.org/goleak#VerifyTestMain ?

Adding that to the TestMain of the `agent` and `consul/agent` packages seems like it would catch not only leaks in production code, but also leaks in tests. There's an option to ignore some functions if we need to ignore freeport.

Reply:
My first pass at this function used ubers `goleak` but it wasn't flexibile enough to detect just the go routines that this test leaked. There are 3 or 4 libraries that we pull in which spawn go routines in an `init` function (not to mention our own freeport which spawns one on first use). `goleak` would always report these extra go routines as leaks.

This test is definitely not comprehensive. The commit message does note that: [`a8605e5` (#8184)](https://github.com/hashicorp/consul/pull/8184/commits/a8605e525b653c8486aad295fb91cc173c7055ec)

There are likely to be other go routines leaked on clients or servers in secondary dcs not to mention with regards to enterprise features. I looked at this as more of a starting place to figure out how to test for leaks and not the final solution that can detect leaks anywhere. A majority of our tests that use test agents spin up a single server and run things against it. So my goal here was to ensure that we are not leaking in that case.

Lastly, sitting in a separate package is necessary unless be got rid of every instance of `t.Parallel`. We need to be able to accurately identity the go routines created by this test and not other parallel tests. Putting this in a separate package ensures that it runs within its own OS process as it will have its own test binary.

Reply:
I will have to look into the filtering options. Its possible with tuning things that it might be okay.

Reply:
Do you have the list of libraries that start goroutines in `init` ? I think that would be a good thing to fix as well.

The readme for uber/goleak has an interesting way of isolating the leaks when running for an entire package: https://github.com/uber-go/goleak#determine-source-of-package-leaks

If the goroutines are detected in `TestMain`, wouldn't all the `t.Parallel` tests have finished at that point?

Reply:
Looks like right now my 3 - 4 should actually have been 2 + freeport.

I added this to both agent and agent/consul and am working out the kinks:

```
func TestMain(m *testing.M) {
	goleak.VerifyTestMain(m,
		goleak.IgnoreTopFunction("k8s.io/klog.(*loggingT).flushDaemon"),
		goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"),
		goleak.IgnoreTopFunction("github.com/hashicorp/consul/sdk/freeport.checkFreedPorts")
	)
}
```

Reply:
Assuming that CI is happy enough with this my plan is to remove the first commit in the separate package in favor of this one.

Reply:
So using ubers goleak with configuring the ignores seems to work well enough, I have replaced the custom test func with the test mains.

Reply:
Well that is a big failure. Tons of other leaks when I enabled it in TestMain.

Reply:
Oh ya, lots of good finds in there. I see some for the cache fetch we knew about.

Do you want to drop from `TestMain` change from this PR and merge just the fixes for now? We can create an issue for fixing the rest of the leaks and then add the TestMain changes back in ?

Reply:
I fixed it, but swapped in goleak over the other lib and added a big comment about this not being ideal and the reasons why we need to do it like this for now.